adapters/repos/db/queue/queue.go

```diff
@@ -0,0 +1,747 @@
+	}
+	if opt.Scheduler == nil {
+		return nil, errors.New("scheduler is required")
+	}
+	if opt.Dir == "" {
+		return nil, errors.New("dir is required")
+	}
+	if opt.TaskDecoder == nil {
+		return nil, errors.New("task decoder is required")
+	}
+
+	if opt.Logger == nil {
+		opt.Logger = logrus.New()
+	}
+	opt.Logger = opt.Logger.WithField("queue_id", opt.ID)
+	if opt.StaleTimeout <= 0 {
+		opt.StaleTimeout = 5 * time.Second
+	}
+	if opt.ChunkSize <= 0 {
+		opt.ChunkSize = defaultChunkSize
+	}
+
+	q := DiskQueue{
+		id:           opt.ID,
+		scheduler:    opt.Scheduler,
+		dir:          opt.Dir,
+		Logger:       opt.Logger,
+		chunkSize:    opt.ChunkSize,
+		staleTimeout: opt.StaleTimeout,
+		taskDecoder:  opt.TaskDecoder,
+		w:            bufio.NewWriterSize(nil, 64*1024),
+		chunkRead:    make(map[string]struct{}),
+	}
+
+	// create the directory if it doesn't exist
+	err := os.MkdirAll(q.dir, 0755)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create directory")
+	}
+
+	// determine the number of records stored on disk
+	q.recordCount, err = q.calculateRecordCount()
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure the partial chunk is created or loaded
+	err = q.ensureChunk()
+	if err != nil {
+		return nil, err
+	}
+
+	// set the last push time to now
+	now := time.Now()
+	q.lastPushTime.Store(&now)
+
+	return &q, nil
+}
+
+// Close the queue, prevent further pushes and unregister it from the scheduler.
+func (q *DiskQueue) Close() error {
+	if q.closed.Swap(true) {
+		return errors.New("queue already closed")
+	}
+
+	q.scheduler.UnregisterQueue(q.id)
+
+	err := q.Flush()
+	if err != nil {
+		return err
+	}
+
+	q.m.Lock()
+	defer q.m.Unlock()
+
+	if q.f != nil {
+		err := q.f.Close()
+		if err != nil {
+			return errors.Wrap(err, "failed to close chunk")
+		}
+
+		q.f = nil
+	}
+
+	return nil
+}
+
+func (q *DiskQueue) ID() string {
+	return q.id
+}
+
+func (q *DiskQueue) Push(record []byte) error {
+	if q.closed.Load() {
+		return errors.New("queue closed")
+	}
+
+	if len(record) == 0 {
+		return errors.New("empty record")
+	}
+
+	now := time.Now()
+	q.lastPushTime.Store(&now)
+
+	q.m.Lock()
+	defer q.m.Unlock()
+
+	err := q.ensureChunk()
+	if err != nil {
+		return err
+	}
+
+	// length of the record in 4 bytes
+	err = binary.Write(q.w, binary.BigEndian, uint32(len(record)))
+	if err != nil {
+		return errors.Wrap(err, "failed to write record length")
+	}
+
+	// write the record
+	n, err := q.w.Write(record)
+	if err != nil {
+		return errors.Wrap(err, "failed to write record")
+	}
+
+	q.partialChunkSize += 4 + uint64(n)
+	q.partialChunkRecordCount++
+	q.recordCount++
+
+	return nil
+}
+
+func (q *DiskQueue) ensureChunk() error {
+	if q.f != nil && q.partialChunkSize >= q.chunkSize {
+		err := q.promoteChunkNoLock()
+		if err != nil {
+			return err
+		}
+	}
+
+	if q.f == nil {
+		// create or open partial chunk
+		path := filepath.Join(q.dir, partialChunkFile)
+		f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY, 0644)
```

Comment:
it may be convenient to also add `os. O_APPEND`

Reply:
ah just saw below that it's not treated as append-only file

Reply:
Exactly, I started with O_APPEND, but I need to update the header before promoting the chunk so it doesn't work.